tests/test_pre_shape.py

```diff
@@ -0,0 +1,239 @@
+        transposed_point = Matrices.transpose(point)
+
+        tmp_expected = gs.matmul(tan, transposed_point)
+        expected = tmp_expected - Matrices.transpose(tmp_expected)
+
+        tmp_result = gs.matmul(vertical, transposed_point)
+        result = tmp_result - Matrices.transpose(tmp_result)
+        self.assertAllClose(result, expected)
+
+    @geomstats.tests.np_and_pytorch_only
+    def test_horizontal_projection(self):
+        w = gs.random.rand(self.m_ambient, self.k_landmarks)
+        point = self.space.random_uniform()
+        tan = self.space.to_tangent(w, point)
+        horizontal = self.space.horizontal_projection(tan, point)
+        transposed_point = Matrices.transpose(point)
+        result = gs.matmul(horizontal, transposed_point)
+        expected = Matrices.transpose(result)
+
+        self.assertAllClose(result, expected)
+
+    @geomstats.tests.np_and_pytorch_only
+    def test_horizontal_projection_vectorized(self):
+        w = gs.random.rand(self.n_samples, self.m_ambient, self.k_landmarks)
+        point = self.space.random_uniform(self.n_samples)
+        tan = self.space.to_tangent(w, point)
+        horizontal = self.space.horizontal_projection(tan, point)
+        transposed_point = Matrices.transpose(point)
+        result = gs.matmul(horizontal, transposed_point)
+        expected = Matrices.transpose(result)
+
+        self.assertAllClose(result, expected)
+
+    @geomstats.tests.np_and_pytorch_only
+    def test_horizontal_and_is_horizontal(self):
+        w = gs.random.rand(self.m_ambient, self.k_landmarks)
+        point = self.space.random_uniform()
+        tan = self.space.to_tangent(w, point)
+        horizontal = self.space.horizontal_projection(tan, point)
+
+        horizontal = gs.stack([horizontal, w])
+        result = self.space.is_tangent(horizontal, point)
+        expected = gs.array([True, False])
+
+        self.assertAllClose(result, expected)
+
+    def test_realign(self):
+        point, base_point = self.space.random_uniform(2)
+        aligned = self.space.realign(point, base_point)
+        alignment = gs.matmul(aligned, Matrices.transpose(base_point))
+        result = Matrices.is_symmetric(alignment)
+        self.assertTrue(result)
+
+    def test_realign_vectorization(self):
+        base_point = self.space.random_uniform()
+        point = self.space.random_uniform(2)
+        aligned = self.space.realign(point, base_point)
+        alignment = gs.matmul(aligned, Matrices.transpose(base_point))
+        result = Matrices.is_symmetric(alignment)
+        self.assertTrue(gs.all(result))
+
+        base_point = self.space.random_uniform(2)
+        point = self.space.random_uniform()
+        aligned = self.space.realign(point, base_point)
+        alignment = gs.matmul(aligned, Matrices.transpose(base_point))
+        result = Matrices.is_symmetric(alignment)
+        self.assertTrue(gs.all(result))
+
+    def test_inner_product(self):
```

Comment:
This only tests the shape of the inner_product. Maybe call it `test_inner_product_shape` and have another test that tests the result?